File: python/searching/breadth_first_search.py
#Breadth First Search 
#Used to find shortest path in graph

#graph for example
graph = {
  'A' : ['B','C'],
  'B' : ['D', 'E'],
  'C' : ['F','G'],
  'D' : [],
  'E' : ['G'],
  'F' : [],
  'G' : []
}

visited = [] # List to keep track of visited nodes.
queue = []     #Initialize a queue

def breadth_first_search(visited, graph, node):
  visited.append(node)
  queue.append(node)

  while queue:
    s = queue.pop(0) 
    print (s, end = " ") 

    for neighbour in graph[s]:
      if neighbour not in visited:
        visited.append(neighbour)
        queue.append(neighbour)

# Driver Code
breadth_first_search(visited, graph, 'A')  #starting search form 'A'

File: cpp/Graphs/Breadth_First_Search.cpp
/*This a Breadth first search implementation in graphs using C++ */
#include<bits/stdc++.h>
using namespace std;


void Breadth_First_Search(int Graph[][5], int n_nodes,int starting, int visited[])
{   visited[starting]=true;
    list<int> queue; 
    queue.push_back(starting);

    while(!queue.empty()){
        int node=queue.front();
        cout<<node<<" ";
        queue.pop_front();
    for(int vertex=0;vertex<n_nodes;vertex++)
        if(!visited[vertex] && Graph[node][vertex]==1){
            visited[vertex]=true;
            queue.push_back(vertex);
        }
    }
}

int main(){
    //Ajancy matrix for connections between nodes
    int Graph[5][5]={
        {0,1,0,1,0},//0->1,3
        {0,0,1,0,0},//1->2
        {0,0,0,0,1},//2->4
        {0,1,0,0,1},//3->1,4
        {1,0,0,0,0} //4->0
    };//initialize a graph with 5 nodes and null connections
    int visited[5];//initialize a visited array with everything false
    for(int node=0;node<5;node++){
        visited[node]=false;
    }
    
    Breadth_First_Search(Graph,5,4,visited);
    return 0;
}
